proposals/p2200.md

```diff
@@ -0,0 +1,1097 @@
+    value of a template parameter are _template validity dependent_. These
+    expressions can trigger a monomorphization error, but are not considered
+    template dependent for purposes of a containing expression.
+
+Note that an expression may not be template dependent even though it has a
+template-dependent sub-expression. For example, a function may have a value that
+is function dependent, but calling that function only needs the type of the
+function (meaning the function's signature) to not be template dependent.
+
+#### Simple member access
+
+There are three cases when performing unqualified member-name lookup into a
+templated type:
+
+```carbon
+fn F[template T:! I](x: T) {
+  x.G();
+}
+```
+
+-   If generic name lookup would succeed, in this example it would be because
+    `G` is a member of `I`, then the result of name lookup is template validity
+    dependent. This means that template instantiation may fail if `T` has a
+    member `G` different than `T.(I.G)`. Assuming it succceeds, though, it will
+    definitely have meaning determined by `I.G`. There may still be ambiguity
+    making the result dependent if it is not known whether `x` is a type and
+    `I.G` is a method and so has an implicit `me` parameter.
+-   If the member name is not found in the constraint, lookup may still succeed
+    once the type is known, so the result is template dependent.
+-   If the lookup is ambiguous prior to knowing the value of the type, for
+    example if `G` has two distinct meanings in `I`, then the code is invalid.
+
+The value of the expression will be dependent. For example, if `U` is an
+associated type of `I`, then `T.U` as an expression is template validity
+dependent, but the value of that expression is dependent. The value is not
+always needed to perform checking, for example `x.G()` can be checked without
+ever determining the value of `x.G` as long as its signature can be determined,
+if it is valid.
+
+#### Compound member access
+
+Adding qualifier to the member name, as in `x.(U.V)`, can make the lookup less
+dependent on the template parameter:
+
+-   If `x` is dependent, and `U` is an interface, the value of the expression is
+    dependent, but the type of the expression is not dependent unless the type
+    of `U.V` involves `Self`. The lookup itself follows proposal
+    [#2360](https://github.com/carbon-language/carbon-lang/pull/2360):
+
+    -   If `U.V` is an _instance_ member, and the type of `x` is known to
+        implement `U`, then the lookup is not dependent. For example, there
+        could be a requirement on `x` or a sufficiently general implementation
+        of `U` that includes all possible types of `x`. The resulting value is
+        dependent unless the implementation of `U` is `final`, see
+        [the impl lookup section](#impl-lookup).
+    -   Otherwise, the lookup is template validity dependent.
+
+    ```carbon
+    interface Hashable {
```

Comment:
```suggestion
    interface Serializable {
      fn Serialize[me: Self]();
    }
    
    interface Printable {
      fn Print[me: Self]();
    }

    interface Hashable {
```
Given that the example below depends on whether `Serialize` is an instance member, and also on the return type of `Serialize` (specifically, if it involves `Self`), it seems useful to define `Serializable` here. Similarly for `Printable`.